linkerd/trace-context/src/propagation.rs

```diff
@@ -42,16 +47,111 @@ impl TraceContext {
 }
 
 pub fn unpack_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
-    unpack_grpc_trace_context(request).or_else(|| unpack_http_trace_context(request))
+    // Attempt to parse as w3c first since it's the newest interface in
+    // distributed tracing ecosystem
+    unpack_w3c_trace_context(request)
+        .or_else(|| unpack_grpc_trace_context(request))
+        .or_else(|| unpack_http_trace_context(request))
 }
 
 // Generates a new span id, writes it to the request in the appropriate
 // propagation format and returns the generated span id.
 pub fn increment_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
     match context.propagation {
-        Propagation::Grpc => increment_grpc_span_id(request, context),
-        Propagation::Http => increment_http_span_id(request),
+        Propagation::B3Grpc => increment_grpc_span_id(request, context),
+        Propagation::B3Http => increment_http_span_id(request),
```

Comment:
And we probably want to move these into a submodule similarly.

```suggestion
        Propagation::B3Grpc => b3::increment_grpc_span_id(request, context),
        Propagation::B3Http => b3::increment_http_span_id(request),
```